staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,640 @@
+	for _, queue := range qs.queues {
+		reqs += queue.GetRequestsExecuting()
+		// It might be best to delete this line. If everything is working
+		//  correctly, there will be no waiting packets if reqs < C on current
+		//  line 85; if something is going wrong, it is more accurate to say
+		// that virtual time advanced due to the requests actually executing.
+
+		// reqs += len(queue.GetPackets())
+		if len(queue.GetPackets()) > 0 || queue.GetRequestsExecuting() > 0 {
+			NEQ++
+		}
+	}
+	// no active flows, virtual time does not advance (also avoid div by 0)
+	if NEQ == 0 {
+		return 0
+	}
+	return math.Min(float64(reqs), float64(qs.concurrencyLimit)) / float64(NEQ)
+}
+
+// updateQueueVirStartTime updates the virtual start time for a queue
+// this is done when a new packet is enqueued.  For more info see:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#dispatching
+func (qs *QueueSetImpl) updateQueueVirStartTime(packet *Packet, queue *Queue) {
+	// When a request arrives to an empty queue with no requests executing:
+	// len(queue.GetPackets()) == 1 as enqueue has just happened prior (vs  == 0)
+	if len(queue.GetPackets()) == 1 && queue.GetRequestsExecuting() == 0 {
+		// the queue’s virtual start time is set to the virtual time.
+		queue.SetVirStart(qs.vt)
+	}
+}
+
+// removeQueueAndUpdateIndexes uses reslicing to remove an index from a slice
+//  and then updates the 'Index' field of the queues to be correct
+func removeQueueAndUpdateIndexes(queues []*Queue, index int) []*Queue {
+	removedQueues := removeIndex(queues, index)
+	for i := index; i < len(removedQueues); i++ {
+		removedQueues[i].SetIndex(removedQueues[i].GetIndex() - 1)
+	}
+	return removedQueues
+}
+
+// removeIndex uses reslicing to remove an index from a slice
+func removeIndex(s []*Queue, index int) []*Queue {
+	return append(s[:index], s[index+1:]...)
+}
+
+// FinishPacketAndDequeueWithChannelAsMuchAsPossible is a convenience method which calls finishPacket
+// for a given packet and then dequeues as many packets as possible
+// and updates that packet's channel signifying it is is dequeued
+// this is a callback used for the filter that the QueueSetImpl supports
+func (qs *QueueSetImpl) FinishPacketAndDequeueWithChannelAsMuchAsPossible(pkt *Packet) {
+	qs.LockAndSyncTime()
+	defer qs.lock.Unlock()
+
+	qs.finishPacket(pkt)
+	qs.DequeueWithChannelAsMuchAsPossible()
+}
+
+// FinishPacket is a callback that should be used when a previously dequeud packet
+// has completed it's service.  This callback updates imporatnt state in the
+// QueueSetImpl
+func (qs *QueueSetImpl) finishPacket(p *Packet) {
+
+	S := qs.clk.Since(p.GetStartTime()).Seconds()
+
+	// When a request finishes being served, and the actual service time was S,
+	// the queue’s virtual start time is decremented by G - S.
+	virstart := p.GetQueue().GetVirStart()
+	virstart -= qs.estimatedServiceTime - S
+	p.GetQueue().SetVirStart(virstart)
```

Comment:
These three lines could be collapsed into one:
```
    p.Queue.VirStart -= qs.estimatedServiceTime - S
```
or, if you want to maintain abstration boundaries between QueueSet, Queue, and Packet:
```
    p.GetQueue().SetVirStart(p.GetQueue().GetVirStart() - (qs.estimatedServiceTime - S))
```

There are other similar opportunities for simplification.

Reply:
I've removed all of the Get/Set methods from the interface and collapsed logic where I saw redundancy (similar to what was pointed out here)